README


-------------------------------------------------------------------------------

Mecrisp-Ice - A conveniently enhanced fork of Swapforth by James Bowman

  It makes your Lattice Icestick FPGA development board
  to behave just like a microcontroller.


The original Swapforth and J1a processor implementation can be found here:
http://www.excamera.com/sphinx/article-j1a-swapforth.html

I would not have been possible without the first open-source FPGA toolchain:
http://www.clifford.at/icestorm/

Many, many thanks to James Bowman and the whole Icestorm team !

http://mecrisp.sourceforge.net/
m-atthias@users.sf.net

-------------------------------------------------------------------------------

HIC SUNT DRACONES - HIC SUNT DRACONES - HIC SUNT DRACONES

This is the first experimental release, so beware: Monsters may be inside !

-------------------------------------------------------------------------------

Shortly after Clifford Wolff, Mathias Lasser and Cotton Seed managed to build
the first working completely open-source FPGA toolchain for the iCE40-HX1K,
James Bowman started porting his elegant J1 processor design.

I have been involved in testing and enhanced the resulting monumental work
with stronger optimisations and a handfull of features for convenience.

-------------------------------------------------------------------------------

What you need:

* A Lattice Icestick with an iCE40-HX1K FPGA
    http://www.latticesemi.com/icestick
    
* Or a Nandland Go with iCE40-HX1K FPGA
    https://www.nandland.com/goboard/introduction.html

* Or a Lattice HX8K breakout board with an iCE40-HX8K FPGA
   
* The Iceprog tool to flash the included binary
    http://www.clifford.at/icestorm/

* A terminal, picocom is fine.

-------------------------------------------------------------------------------

Which board for you ?

If you just wish to play with a stack processor,
if you have small ideas,
if you desire to have IrDA communication,
if you wish a cheap, cool looking board,
if you are fine without tinkering with additional logic

  --> HX1K Icestick is your choice.
  
If you wish to experiment with VGA video,
if you like 7-segment displays,
if you need buttons on your board

  --> HX1K Nandland Go should be the right for you.
  
If you have large ideas in Forth,
if you need a lot of IO pins,
if you wish to wire in additional peripherals, opcodes or whatever you imagine,
if you love to have single cycle multiplier and barrel shifters

  --> HX8K breakout board will give you a nice time.  

-------------------------------------------------------------------------------

Basic usage, same for HX1K and HX8K:

  Flash the binary:
  iceprog -b j1a.bin

  Open a terminal:
  picocom -b 115200 /dev/ttyUSB$n --imap lfcrlf,crcrlf --omap delbs,crlf --send-cmd "ascii-xfr -s -l 30 -n"

  Toggle DTR line to Reset by pressing Crtl+A Ctrl+T Crtl+A Ctrl+T

  You should get a welcome message.

  If you hard crash it, iceprog -t or a power cycle should recover.

-------------------------------------------------------------------------------

Some notes for Forth users new to this target and its peripheral possibilities:

The Icestick has 8 kb of initialised RAM which is mapped from address
$0000 to $1FFF and a SPI flash to store the FPGA bitstream and additional
user generated memory images.

Data and return stack are 16 elements deep. James implemented them as
part of the CPU itself, so they are not accessible through memory.

Pinout and board layout:

       /--------------------------------------
      /                 Header 1:  76543210-+
     /            ________              ++
    /            |        |      R      --
[USB]  FT        |  FPGA  |    R G R    73   Ir
[USB]  DI        |        |      R      62   DA
    \            |________|             51
     \                                  40
      \                 Header 2:  01234567-+
       \--------------------------------------

  Numbers are bit numbers
  + is 3.3 V   - is GND
  G Green LED  R Red LEDs


There is a separate IO address space, with those registers available:

     Addr  Bit READ            WRITE

     0001  0   PMOD in
     0002  1   PMOD out        PMOD out
     0004  2   PMOD dir        PMOD dir
     0008  3   misc.out        misc.out

     0010  4   header 1 in
     0020  5   header 1 out    header 1 out
     0040  6   header 1 dir    header 1 dir
     0080  7                   eint

     0100  8   header 2 in
     0200  9   header 2 out    header 2 out
     0400  10  header 2 dir    header 2 dir
     0800  11                  dint

     1000  12  UART RX         UART TX
     2000  13  misc.in
     4000  14  ticks           clear ticks
     8000  15

Contents of misc.out and misc.in:

   Bitmask Bit  misc.out        misc.in

     0001    0  SPI CS          UART Ready to Transmit
     0002    1  SPI MOSI        UART Character received
     0004    2  SPI SCK         SPI MISO
     0008    3  IrDA-TXD        IrDA-RXD
     0010    4  IrDA-Sleep      RTS
     0020    5  CTS             Random
     0040    6  Green LED
     0080    7  Red LED
     0100    8  Red LED
     0200    9  Red LED
     0400   10  Red LED

If you are already familiar with MSP430 chips,
you will know this IO port style:

* The IN register gives the current electrical state
* The contents of the OUT registers determine what level the outputs should be
* The DIR register let you switch a pin to be an output by writing an one into

You can set two registers at once if you OR together their addresses.
  255 $440 io! should set all header pins as outputs.

Inputs can ORed together the same way and give an ORed result.

You can detect short-circuited pins -
  for example if a pin is set to output and low, but shortened to Vcc,
  then the OUT register will read back low, as set by you, but the
  IN register will read high for that pin.

Misc.out and misc.in are a mixed back of wires and flags available.

No fear to destroy the IrDA transceiver with too long pulses as it turns off
the IR LED itself when IrDA-TXD is high for more than 50 us.

The UART data register is for both incoming and outgoing data,
a read from it will clear the "Character received" flag
and you should only write to it when "Ready to Transmit" is set.

Ticks contains a 16 bit cycle counter that counts up with 48 MHz and
you can set it to any value you wish by writing to it.

Memory location $1FFE is an interrupt vector for the ticks counter overflow.
You can place an opcode there, perhaps ALU Exit ($608C) or a JMP to a handler.
Interrupts can be enabled with eint and disabled with dint.

-------------------------------------------------------------------------------

Notes on HX8K:
--------------

This feels very similiar to Mecrisp-Ice on HX1K, just with a few differences:

  + Double amount of memory: 16 kb instead of 8 kb
  + Double stack depth: 32 elements each instead of 16
  + Barrel shifter for single cycle lshift rshift arshift
  + Single cycle multiplication
  + Fast opcodes for 1+ and 1-
  + rdepth
  + 8 LEDs instead of 5
  + Lots of gates free for experiments
  - 36 MHz only
  - No IrDA
  - Variable @ needs 2 instructions and 5 cycles instead of 1 instruction and 2 cycles  

The peripheral map is the same, but the IN, OUT and DIR registers
are up to 16 bits wide, with the following connections:

 $0001 Port A IN     $0010 Port B IN     $0100 Port C IN 
 $0002 Port A OUT    $0020 Port B OUT    $0200 Port C OUT
 $0004 Port A DIR    $0040 Port B DIR    $0400 Port C DIR

 
                     Red LEDs
                     76543210
  1.2V 3.3V   oo                 oo   oo
  1.2V B13    oo                 oo   oo
  A12  B12    oo     ________    oo   oo
  GND  GND    oo    |        |   oo   oo
  A11  B11    oo    |  FPGA  |   oo   oo
  A10  B10    oo    |        |   oo   oo
  A9   B9     oo    |________|   oo   oo
  GND  GND    oo                 oo   oo
  A8   B8     oo                 oo   oo
  A7   B7     oo                 oo   oo
  A6   B6     oo  Jumper config: oo   oo
  GND  GND    oo                 oo   oo
  A5   B5     oo     |     ||    oo   oo
  A4   B4     oo                 oo   oo
  A3   B3     oo        Green    oo   oo
  GND  GND    oo        LED      oo   oo
  A2   B2     oo     FT          oo   oo
  A1   B1     oo     DI          oo   oo
  A0   B0     oo                 oo   oo
  GND  GND    oo     [USB]       oo   oo
                     [USB]
  

  Port C is currently unconnected, you can wire the bits as you wish
  in the constraints PCF file.
  
Ticks contains a 16 bit cycle counter that counts up with 36 MHz and
you can set it to any value you wish by writing to it.

Memory location $3FFE is an interrupt vector for the ticks counter overflow.
You can place an opcode there, perhaps ALU Exit ($608C) or a JMP to a handler.
Interrupts can be enabled with eint and disabled with dint.  
  
-------------------------------------------------------------------------------

Notes on Nandland Go:
---------------------

S4 is Reset button, as DTR line is not connected on this board.

     Addr  Bit READ            WRITE

     0001  0   PMOD in
     0002  1   PMOD out        PMOD out
     0004  2   PMOD dir        PMOD dir
     0008  3   misc.out        misc.out

     0010  4                   VGA Colours
     0020  5                   VGA HSYNC
     0040  6                   VGA VSYNC
     0080  7   Segments        Segments

     0100  8
     0200  9
     0400  10
     0800  11

     1000  12  UART RX         UART TX
     2000  13  misc.in
     4000  14  ticks           clear ticks
     8000  15

Contents of misc.out and misc.in:

   Bitmask Bit  misc.out        misc.in
 
     0001    0  SPI CS          UART Ready to Transmit
     0002    1  SPI MOSI        UART Character received
     0004    2  SPI SCK         SPI MISO
     0008    3  Green LED 1     Random
     0010    4  Green LED 2     S1
     0020    5  Green LED 3     S2
     0040    6  Green LED 4     S3 
     
Note that you can OR together the VGA lines.

    VSYNC HSYNC Blue 2 Blue 1 Blue 0 Green 2 Green 1 Green 0 Red 2 Red 1 Red 0

$10:               256    128     64      32      16       8     4     2     1
$20:        512
$40: 1024

So by writing IO Register $60, you can set HSYNC and VSYNC independently
from the colour register bits. If you prefer to have all VGA signals
in one register, use IO port $70.

Ticks contains a 16 bit cycle counter that counts up with 25 MHz and
you can set it to any value you wish by writing to it.

Memory location $1FFE is an interrupt vector for the ticks counter overflow.
You can place an opcode there, perhaps ALU Exit ($608C) or a JMP to a handler.
Interrupts can be enabled with eint and disabled with dint.
  
-------------------------------------------------------------------------------

Special features included in Mecrisp-Ice:
-----------------------------------------

You can save the complete current memory contents into the SPI flash
sectors 1 to 63. Sector 0 contains the FPGA bitstream, sector 1 will be
loaded automatically on startup if you save a memory image into this.

sector# load
sector# save
sector# erase

There is a variable "init" which is executed after loading an image.

So if you wish to create a turnkey application, prepare everything in RAM
and type:

  ' fireworks init !
  1 save

Replace fireworks with whatever you wish to be executed on startup.
Nothing is executed if init contains a zero, which is the default.

If you are not sure if it will work properly, save to any other sector
and load manually. The init mechanism works the same on all sectors.

On HX8K: Sectors 0, 1 and 2 contain the bitstream.
         Sector 3 will be loaded automatically if available.

The Nandland Go has a smaller SPI Flash with four sectors only:

  0: Bitstream
  1: Autoload
  2:  Free
  3:  Free         
         
-----------------------------------------

For your convenience, some introspection features are preloaded for default.

Things to try:

  0 8192 dump
  insight
  see +

They will be removed by typing "new" so that you have more space available.

-----------------------------------------

Inline optimisations and constant folding are included !

-------------------------------------------------------------------------------

Main differences to Mecrisp and Mecrisp-Stellaris:

Unlike Mecrisp and Mecrisp-Stellaris, this is running completely
in RAM memory. So there is no need for special memory handling !

You can use CREATE with DOES> and if you need a buffer,
just CREATE ... ALLOT it.

By the way: If you save a memory image, the last contents of the variables
are saved during the process and reloaded with the image. Variables are
initialised, but there is no logic to reinitialise them on Reset to the
values assigned at compilation.

-------------------------------------------------------------------------------

If you wish to build the whole package yourself:

* Gforth
* Freepascal
* Yosys for synthesis
* Arachne-PNR for place and route
* Icestorm tools for bitstream generation

... and their dependencies.

In a nutshell, you need to type:

  compile
  iceprog -b j1a.bin

That's all.

;------------------------------------------------------------------------------
Here comes a word list,                                          -- Glossary --
  with short descriptions of all currently included words:
View it with fixed-width font !
;------------------------------------------------------------------------------

;------------------------------------------------------------------------------
; Terminal-IO
;------------------------------------------------------------------------------

        emit?           ( -- Flag ) Ready to send a character ?
        key?            ( -- Flag ) Checks if a key is waiting
        key             ( -- Char ) Waits for and fetches the pressed key
        emit            ( Char -- ) Emits a character.

;------------------------------------------------------------------------------
; Stack Jugglers
;------------------------------------------------------------------------------

Single-Jugglers:

        depth           ( -- +n ) Gives number of single-cell stack items.
        rdepth          ( -- +n ) HX8K only: Number of items on return stack.
        nip             ( x1 x2 -- x2 )
        drop            ( x -- )
        rot             ( x1 x2 x3 -- x2 x3 x1 )
        -rot            ( x1 x2 x3 -- x3 x1 x2 )
        swap            ( x1 x2 -- x2 x1 )
        tuck            ( x1 x2 -- x2 x1 x2 )
        over            ( x1 x2 -- x1 x2 x1 )
        ?dup            ( x -- 0 | x x )
        dup             ( x -- x x )

        >r              ( x -- ) (R: -- x )
        r>              ( -- x ) (R: x -- )
        r@              ( -- x ) (R: x -- x )
        rdrop           (  --  ) (R: x -- )

Double-Jugglers:        They perform the same for double numbers.

        2drop           ( x1 x2 -- )
        2swap           ( x1 x2 x3 x4 -- x3 x4 x1 x2 )
        2over           ( x1 x2 x3 x4 -- x1 x2 x3 x4 x1 x2 )
        2dup            ( x1 x2 -- x1 x2 x1 x2 )

;------------------------------------------------------------------------------
; Logic
;------------------------------------------------------------------------------

        arshift         ( x1 u -- x2 ) Arithmetric right-shift of u bit-places
        rshift          ( x1 u -- x2 ) Logical right-shift of u bit-places
        lshift          ( x1 u -- x2 ) Logical  left-shift of u bit-places
        invert          ( x1 -- x2 )   Invert all bits
        not             ( x1 -- x2 )   Invert all bits = Bitwise not
        xor             ( x1 x2 -- x3 ) Bitwise Exclusive-OR
        or              ( x1 x2 -- x3 ) Bitwise OR
        and             ( x1 x2 -- x3 ) Bitwise AND
        false           ( --  0 ) False-Flag
        true            ( -- -1 ) True-Flag

;------------------------------------------------------------------------------
; Arithmetic for single numbers
;------------------------------------------------------------------------------

        /mod            ( n1 n2 -- n3 n4 ) n1 / n2 = n4 rem n3
        mod             ( n1 n2 -- n3 ) n1 / n2 = remainder n3
        /               ( n1 n2 -- n3 ) n1 / n2 = n3
        *               ( u1|n1 u2|n2 -- u3|n3 ) 16*16 = 16 Multiplication
        min             ( n1 n2 -- n1|n2 ) Keeps smaller of top two items
        max             ( n1 n2 -- n1|n2 ) Keeps greater of top two items
        umin            ( u1 u2 -- u1|u2 ) Keeps unsigned smaller
        umax            ( u1 u2 -- u1|u2 ) Keeps unsigned greater
        1-              ( u1|n1 -- u2|n2 ) Subtracts one, optimized
        1+              ( u1|n1 -- u2|n2 ) Adds one, optimized

        2*              ( n1 -- n2 ) Arithmetric  left-shift
        2/              ( n1 -- n2 ) Arithmetric right-shift
        abs             ( n -- u ) Absolute value
        negate          ( n1 -- n2 ) Negate
        sgn             ( u1 n1 -- n2 ) Give u1 the sign of n2
        -               ( u1|n1 u2|n2 -- u3|n3 ) Subtraction
        +               ( u1|n1 u2|n2 -- u3|n3 ) Addition

;------------------------------------------------------------------------------
; Arithmetic involving double numbers
;------------------------------------------------------------------------------

        um*             ( u1 u2 -- ud )    16*16 = 32 Multiplication

        um/mod          ( ud u1 -- u2 u3 ) ud / u1 = u3 remainder u2

        m+              ( d1 n -- d2 ) Addition of a double with a single
        m*              ( n1 n2 -- d )     n1 * n2 = d
        fm/mod          ( d n1 -- n2 n3 )  d / n1 = n3 remainder r2 floored
        sm/rem          ( d n1 -- n2 n3 )  d / n1 = n3 remainder r2 symmetric

        */              ( n1 n2 n3 -- n4 )     n1 * n2 / n3 = n4
        */mod           ( n1 n2 n3 -- n4 n5 )  n1 * n2 / n3 = n5 remainder n4

        d2*             ( d1 -- d2 ) Arithmetric  left-shift

        dabs            ( d -- ud ) Absolute value
        dnegate         ( d1 -- d2 ) Negate
        d-              ( ud1|d1 ud2|d2 -- ud3|d3 ) Subtraction
        d+              ( ud1|d1 ud2|d2 -- ud3|d3 ) Addition
        s>d             ( n -- d ) Makes a signed single number double length

;------------------------------------------------------------------------------
; Comparisions
;------------------------------------------------------------------------------

        u<=             ( u1 u2 -- flag )   Unsigned comparisions
        u>=             ( u1 u2 -- flag )
        u>              ( u1 u2 -- flag )
        u<              ( u1 u2 -- flag )
        <=              ( n1 n2 -- flag )   Signed comparisions
        >=              ( n1 n2 -- flag )
        >               ( n1 n2 -- flag )
        <               ( n1 n2 -- flag )
        0>              ( n -- flag )       Positive ?
        0<              ( n -- flag )       Negative ?
        0<>             ( x -- flag )
        0=              ( x -- flag )
        <>              ( x1 x2 -- flag )
        =               ( x1 x2 -- flag )

        d0=             ( d -- flag )

        within          ( x1 x2 x3 -- ? )   Check if x1 is within x2 and x3.

;------------------------------------------------------------------------------
; Tools for number input
;------------------------------------------------------------------------------

        number          ( c-addr len -- 0 )
                                     -- n 1 )
                                     -- n-low n-high 2 )
                        Tries to convert a string to a number.

;------------------------------------------------------------------------------
; Number base
;------------------------------------------------------------------------------

        binary          ( -- ) Sets base to 2
        decimal         ( -- ) Sets base to 10
        hex             ( -- ) Sets base to 16
        base            ( -- a-addr ) Base variable address

;------------------------------------------------------------------------------
; Memory access
;------------------------------------------------------------------------------

        cmove           ( c-addr1 c-addr2 u -- ) Moves backward
        cmove>          ( c-addr1 c-addr2 u -- ) Moves forward
        move            ( c-addr1 c-addr2 u -- ) Moves u Bytes in Memory
        fill            ( c-addr u c ) Fill u Bytes of Memory with value c

        constant  name  ( u|n -- )  Makes a single constant.
        variable  name  ( u|n -- )  Makes an initialized single variable

        2@              ( a-addr -- ud|d ) Fetches double number from memory
        2!              ( ud|d a-addr -- ) Stores double number in memory

        @               ( a-addr -- u|n ) Fetches single number from memory
        !               ( u|n a-addr -- ) Stores single number in memory
        +!              ( u|n a-addr -- ) Add to memory location

        c@              ( c-addr -- char ) Fetches byte from memory
        c!              ( char c-addr ) Stores byte in memory

IO memory area:

        io@             ( c-addr -- x ) Fetches from IO register
        io!             ( x c-addr -- ) Stores  into IO register

        xor!            ( mask c-addr -- ) Toggle bits
        bic!            ( mask c-addr -- ) Clear BIts
        bis!            ( mask c-addr -- ) Set BIts

;------------------------------------------------------------------------------
; Strings and beautiful output
;------------------------------------------------------------------------------

String routines:

        type            ( c-addr length -- )
                        Prints a string.

        rtype           ( c-addr length u -- )
                        Prints a string in a field u characters wide.

        s" Hello"       Compiles a string and
                        ( -- c-addr length )
                        gives back its address and length when executed.

        ." Hello"       Compiles a string and
                        ( -- )
                        prints it when executed.

        ( Comment )     Ignore Comment
        \ Comment       Comment to end of line

        cr              ( -- ) Emits line feed
        bl              ( -- 32 ) ASCII code for Space
        space           ( -- ) Emits space
        spaces          ( n -- ) Emits n spaces if n is positive

        accept          ( c-addr maxlength -- length ) Read input into a string.

Counted string routines:

        count           ( cstr-addr -- c-addr length )
                        Convert counted string into addr-length string

Pictured numerical output:

        [char] *        Compiles code of following char
                        ( -- char ) when executed

        char *          ( -- char ) gives code of following char
        hold            ( char -- ) Adds character to pictured number
                                    output buffer from the front.

        sign            ( n -- ) Add a minus sign to pictured number
                                 output buffer, if n is negative

        #S              ( ud1|d1 -- 0 0 ) Add all remaining digits
                        from the double length number to output buffer
        #               ( ud1|d1 -- ud2|d2 ) Add one digit from the
                        double length number to output buffer
        #>              ( ud|d -- c-addr len )
                        Drops double-length number and finishes
                        pictured numeric output ready for type
        <#              ( -- ) Prepare pictured number output buffer
        u.              ( u -- ) Print unsigned single number
        .               ( n -- ) Print single number
        ud.             ( ud -- ) Print unsigned double number
        d.              ( d -- ) Print double number


CHANGELOG

Original work:
J1a processor and Swapforth by James Bowman

19. September 2015: Mecrisp-Ice 0.1

Equipped by Matthias Koch with:

* MSP430 style IO
* Ticks counter
* Inline optimisations
* Constant folding
* Load & Save from and to SPI flash with init mechanism
* More readable disassembler
* Dictionary structure insight

* Ledcomm
* IrDA example

22. September 2015: Mecrisp-Ice 0.2

* .s will not crash anymore on deep stacks
* Disassembler improved

27. September 2015: Mecrisp-Ice 0.3

* Clock speed increased to 48 MHz
* Added randombit and random which use clock jitter
  from a free running ring oscillator

29. February 2016: Mecrisp-Ice 0.4

* Minus opcode added to CPU
* Small speed improvements
* Ice Cream Machine emulates J1a and helps building the package in one step
* Removed tethered mode

07. March 2016: Mecrisp-Ice 0.5

* Single cycle io@
* Ticks counter overflow interrupt, $1FFE is vector. Place exit or a JMP opcode there.
* Added eint and dint
* Removed SB_Warmboot

20. April 2016: Mecrisp-Ice 0.6

* Fixed faulty logic for interrupts which happen in a memory fetch cycle.

23. April 2016: Mecrisp-Ice 0.7

* New binary image for HX8K !

  + Double amount of memory: 16 kb instead of 8 kb
  + Double stack depth: 32 elements each instead of 16
  + Barrel shifter for single cycle lshift rshift arshift
  + Single cycle multiplication
  + Fast opcodes for 1+ and 1-
  + rdepth
  + 8 LEDs instead of 5
  + Lots of gates free for experiments
  - 36 MHz only
  - No IrDA
  - Variable @ needs 2 instructions and 5 cycles instead of 1 instruction and 2 cycles

* Reentrant um* and do..loop for interrupt usage

* Larger address range made a different dictionary header structure necessary
  to still accomodate flags for foldability and immediate.
  
* Tidied up nucleus Forth source code  

* Improved compatibility with Mecrisp for MSP430:

  * Initialised variables
  * LF only instead of CR+LF
  
05. June 2016, Mecrisp-Ice 0.8

* New target: Nandland Go board by Russell Merrick
* Timer can be set to any value now
* Improved interrupt handling, with dint and eint as opcodes
* Added eint?



nandland/icestorm/j1a.v
`timescale 1 ns / 1 ps

`default_nettype none
`define WIDTH 16

module SB_RAM2048x2(
	output  [1:0] RDATA,
	input         RCLK, RCLKE, RE,
	input  [10:0] RADDR,
	input         WCLK, WCLKE, WE,
	input  [10:0] WADDR,
	input   [1:0] MASK, WDATA
);
	parameter INIT_0 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_1 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_2 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_3 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_4 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_5 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_6 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_7 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_8 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_9 = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_A = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_B = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_C = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_D = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_E = 256'h0000000000000000000000000000000000000000000000000000000000000000;
	parameter INIT_F = 256'h0000000000000000000000000000000000000000000000000000000000000000;

  wire [15:0] rd;

  SB_RAM40_4K #(
    .WRITE_MODE(3),
    .READ_MODE(3),
    .INIT_0(INIT_0),
    .INIT_1(INIT_1),
    .INIT_2(INIT_2),
    .INIT_3(INIT_3),
    .INIT_4(INIT_4),
    .INIT_5(INIT_5),
    .INIT_6(INIT_6),
    .INIT_7(INIT_7),
    .INIT_8(INIT_8),
    .INIT_9(INIT_9),
    .INIT_A(INIT_A),
    .INIT_B(INIT_B),
    .INIT_C(INIT_C),
    .INIT_D(INIT_D),
    .INIT_E(INIT_E),
    .INIT_F(INIT_F)
  ) _ram (
    .RDATA(rd),
    .RADDR(RADDR),
    .RCLK(RCLK), .RCLKE(RCLKE), .RE(RE),
    .WCLK(WCLK), .WCLKE(WCLKE), .WE(WE),
    .WADDR(WADDR),
    .MASK(16'h0000), .WDATA({4'b0, WDATA[1], 7'b0, WDATA[0], 3'b0}));

  assign RDATA[0] = rd[3];
  assign RDATA[1] = rd[11];

endmodule

module top(input oscillator,

           output D1, 
           output D2, 
           output D3, 
           output D4,
           
           input S1,
           input S2,
           input S3,
           input S4,           

           output TXD,        // UART TX
           input  RXD,        // UART RX

           output SCK,        // Flash SCK
           input  MISO,      // Flash MISO
           output MOSI,     // Flash MOSI
           output CS,      // Flash CS

           inout PORTA0,
           inout PORTA1,
           inout PORTA2,
           inout PORTA3,
           inout PORTA4,
           inout PORTA5,
           inout PORTA6,
           inout PORTA7,

           output Segment1A,
           output Segment1B,
           output Segment1C,
           output Segment1D,
           output Segment1E,
           output Segment1F,
           output Segment1G,
           output Segment2A,
           output Segment2B,
           output Segment2C,
           output Segment2D,
           output Segment2E,
           output Segment2F,
           output Segment2G,
           
           output VGA_HSync,
           output VGA_VSync,

           output VGA_Red_0,
           output VGA_Red_1,
           output VGA_Red_2,
           output VGA_Grn_0,
           output VGA_Grn_1,
           output VGA_Grn_2,
           output VGA_Blu_0,
           output VGA_Blu_1,
           output VGA_Blu_2
);

  // wire resetq = 1'b1;
  wire resetq = ~S4;

  wire clk = oscillator;

  wire io_rd, io_wr;
  wire [15:0] mem_addr;
  wire mem_wr;
  wire [15:0] dout;
  wire [15:0] io_din;
  wire [12:0] code_addr;

  reg unlocked = 0;

`include "../build/ram.v"

  reg interrupt = 0;

  // ######   PROCESSOR   #####################################

  j1 _j1(
    .clk(clk),
    .resetq(resetq),
    .io_rd(io_rd),
    .io_wr(io_wr),
    .mem_wr(mem_wr),
    .dout(dout),
    .io_din(io_din),
    .mem_addr(mem_addr),
    .code_addr(code_addr),
    .insn_from_memory(insn),
    .interrupt_request(interrupt)
  );

  // ######   TICKS   #########################################

  reg [15:0] ticks;

  wire [16:0] ticks_plus_1 = ticks + 1;

  always @(posedge clk)
    if (io_wr & mem_addr[14])
      ticks <= dout;
    else
      ticks <= ticks_plus_1;

  always @(posedge clk) // Generate interrupt on ticks overflow
    interrupt <= ticks_plus_1[16];

  // ######   PMOD   ##########################################

  reg [7:0] pmod_dir;   // 1:output, 0:input
  reg [7:0] pmod_out;
  wire [7:0] pmod_in;

  SB_IO #(.PIN_TYPE(6'b1010_01)) io0 (.PACKAGE_PIN(PORTA0), .D_OUT_0(pmod_out[0]), .D_IN_0(pmod_in[0]), .OUTPUT_ENABLE(pmod_dir[0]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io1 (.PACKAGE_PIN(PORTA1), .D_OUT_0(pmod_out[1]), .D_IN_0(pmod_in[1]), .OUTPUT_ENABLE(pmod_dir[1]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io2 (.PACKAGE_PIN(PORTA2), .D_OUT_0(pmod_out[2]), .D_IN_0(pmod_in[2]), .OUTPUT_ENABLE(pmod_dir[2]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io3 (.PACKAGE_PIN(PORTA3), .D_OUT_0(pmod_out[3]), .D_IN_0(pmod_in[3]), .OUTPUT_ENABLE(pmod_dir[3]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io4 (.PACKAGE_PIN(PORTA4), .D_OUT_0(pmod_out[4]), .D_IN_0(pmod_in[4]), .OUTPUT_ENABLE(pmod_dir[4]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io5 (.PACKAGE_PIN(PORTA5), .D_OUT_0(pmod_out[5]), .D_IN_0(pmod_in[5]), .OUTPUT_ENABLE(pmod_dir[5]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io6 (.PACKAGE_PIN(PORTA6), .D_OUT_0(pmod_out[6]), .D_IN_0(pmod_in[6]), .OUTPUT_ENABLE(pmod_dir[6]));
  SB_IO #(.PIN_TYPE(6'b1010_01)) io7 (.PACKAGE_PIN(PORTA7), .D_OUT_0(pmod_out[7]), .D_IN_0(pmod_in[7]), .OUTPUT_ENABLE(pmod_dir[7]));

  // ######   UART   ##########################################

  wire uart0_valid, uart0_busy;
  wire [7:0] uart0_data;
  wire uart0_wr = io_wr & mem_addr[12];
  wire uart0_rd = io_rd & mem_addr[12];
  wire UART0_RX;
  buart _uart0 (
     .clk(clk),
     .resetq(1'b1),
     .rx(RXD),
     .tx(TXD),
     .rd(uart0_rd),
     .wr(uart0_wr),
     .valid(uart0_valid),
     .busy(uart0_busy),
     .tx_data(dout[7:0]),
     .rx_data(uart0_data));

  // ######   LEDS & PIOS   ###################################

  reg [2:0] PIOS;
  assign {SCK, MOSI, CS} = PIOS;
  reg [3:0] LEDS;
  assign {D4,D3,D2,D1} = LEDS;

  reg [13:0] SEGMENTS;
  assign {Segment2G, Segment2F, Segment2E, Segment2D, Segment2C, Segment2B, Segment2A,
          Segment1G, Segment1F, Segment1E, Segment1D, Segment1C, Segment1B, Segment1A } = ~SEGMENTS;
          
  reg [10:0] VGA;
  assign { VGA_VSync, VGA_HSync,
           VGA_Blu_2, VGA_Blu_1, VGA_Blu_0,
           VGA_Grn_2, VGA_Grn_1, VGA_Grn_0,
           VGA_Red_2, VGA_Red_1, VGA_Red_0
         } = VGA;
  
  // ######   RING OSCILLATOR   ###############################

  wire [1:0] buffers_in, buffers_out;
  assign buffers_in = {buffers_out[0:0], ~buffers_out[1]};
  SB_LUT4 #(
          .LUT_INIT(16'd2)
  ) buffers [1:0] (
          .O(buffers_out),
          .I0(buffers_in),
          .I1(1'b0),
          .I2(1'b0),
          .I3(1'b0)
  );

  wire random = ~buffers_out[1];

  // ######   IO PORTS   ######################################

  /*        bit READ            WRITE

      0001  0   PMOD in
      0002  1   PMOD out        PMOD out
      0004  2   PMOD dir        PMOD dir
      0008  3   misc.out        misc.out

      0010  4                   VGA Colours
      0020  5                   VGA HSYNC
      0040  6                   VGA VSYNC
      0080  7   Segments        Segments

      0100  8 
      0200  9 
      0400  10
      0800  11

      1000  12  UART RX         UART TX
      2000  13  misc.in
      4000  14  ticks           clear ticks
      8000  15
  */

  assign io_din =

    (mem_addr[ 0] ? { 8'd0, pmod_in}                                                 : 16'd0) |
    (mem_addr[ 1] ? { 8'd0, pmod_out}                                                : 16'd0) |
    (mem_addr[ 2] ? { 8'd0, pmod_dir}                                                : 16'd0) |
    (mem_addr[ 3] ? { 9'd0, LEDS, PIOS}                                              : 16'd0) |

    (mem_addr[ 7] ? { 2'd0, SEGMENTS}                                                : 16'd0) |

    (mem_addr[12] ? { 8'd0, uart0_data}                                              : 16'd0) |
    (mem_addr[13] ? { 9'd0, S3, S2, S1, random, MISO, uart0_valid, !uart0_busy}      : 16'd0) |
    (mem_addr[14] ?         ticks                                                    : 16'd0) ;

  // Very few gates needed: Simply trigger warmboot by any IO access to $8000 / $8001 / $8002 / $8003.
  // SB_WARMBOOT _sb_warmboot ( .BOOT(io_wr & mem_addr[15]), .S1(mem_addr[1]), .S0(mem_addr[0]) );

  always @(posedge clk) begin

    if (io_wr & mem_addr[1])  pmod_out <= dout[7:0];
    if (io_wr & mem_addr[2])  pmod_dir <= dout[7:0];
    if (io_wr & mem_addr[3])  {LEDS, PIOS} <= dout[6:0];
    
    if (io_wr & mem_addr[4])  {VGA[8:0]}   <= dout[8:0];
    if (io_wr & mem_addr[5])  {VGA[  9]}   <= dout[  9];
    if (io_wr & mem_addr[6])  {VGA[ 10]}   <= dout[ 10];
    
    if (io_wr & mem_addr[7])  SEGMENTS <= dout[13:0];
  end

  // ######   MEMLOCK   #######################################

  // This is a workaround to protect memory contents during Reset.
  // Somehow it happens sometimes that the first memory location is corrupted during startup,
  // and as an IO write is one of the earliest things which are done, memory write access is unlocked
  // only after the processor is up and running and sending its welcome message.

  always @(negedge resetq or posedge clk)
  if (!resetq) unlocked <= 0;
  else         unlocked <= unlocked | io_wr;

endmodule // top
